### PR TITLE
Modify nectm api

### PR DESCRIPTION
**Pull request by higherstar**
Issue https://github.com/keeleinstituut/tv-tolkevarav/issues/232

### CODE REVIEW

**Review by DanBondarenko — CHANGES_REQUESTED** [review]
Hi @higherstar! Thank you for the pull request.

I clarified the scope of https://github.com/keeleinstituut/tv-tolkevarav/issues/232 with Markus and Marius. We found that further work is still needed to satisfy the requirements. In order to consider the issue completed, the following must be satisfied:
* **Checking whether the JWT is valid should happen in ALL endpoints.** Ideally, there would be some middleware-like code, which is automatically applied to all requests. I suspect Flask might have a solution for this. 
  * The following should be checked: (1) JWT is present in request, (2) JWT is signed using the public key of our SSO, (3) JWT contains user ID and the selected institution ID (both nested under the `"tolkevarav"` claim). 
  * If an incoming request to any endpoint _does not_ satisfy these requirements, it should be rejected.
* **User in institution A must never be able to read OR write data of institution B.** This applies to ALL endpoints. 
  * I can see that this requirement is not easy – there are many endpoints and the logic of how data is accessed/manipulated varies between them. 
  * Ideally, we would define this in a way that avoids duplication – a solution that would be  reusable across resources and endpoints. Since you have been working with NecTM for the last few weeks, you have a better understanding of their data structure, so I would like you to propose a solution for this. 
* **Note that there are some parts of the NecTM codebase which don’t need to be updated – because we will not use it in our system (at least for the current phase).**
  * This includes: (1) user management, (2) authentication endpoints, (3) token endpoint, (4) user settings management and probably some more (I’m not so familiar with the code as to know for certain at this point).
  * I would recommend commenting out the declarations of these "resource endpoints" in `src/RestApi/Api.py`.

And as always, please feel 110% free to PM me or call me at any point – I’m more than happy to discuss anything or answer questions. It’s in the interest of the project that we have clear and thorough communication. :)

File: src/RestApi/Auth.py [hunk 1]
```diff
@@ -34,70 +35,64 @@
...
 # User permission requires either admin or user role
 user_permission = Permission(RoleNeed('user')).union(admin_permission)
 
-def authenticate(username, password):
-  user = Users.query.get(username)
-  if user and user.check_password(password):
-    return user
 
+def current_identity():
+    return Users.query.filter_by(uuid=current_userid()).first()
```
[the remark this answers is not in the document]
Comment (higherstar) [reply src/RestApi/Auth.py:48]: What about User Scope and User Settings Table?

[the remark this answers is not in the document]
Comment (DanBondarenko) [reply src/RestApi/Auth.py:48]: @higherstar User data is stored in another service. All user data need by the translation memory service should be read from the JWT (e.g. privileges, institution, name, IDs etc).

File: src/RestApi/Auth.py [hunk 2]
```diff
@@ -34,70 +35,64 @@
...
+def permission(*roles):
+    def decorator(func):
+        @wraps(func)
+        @oidc.accept_token(require_token=True, render_errors=False)
+        def decorated_function(*args, **kwargs):
+            token_info = g.oidc_token_info
+            if token_info and 'realm_access' in token_info and 'roles' in token_info['realm_access']:
+                user_roles = token_info['realm_access']['roles']
+                for role in roles:
+                    if role not in user_roles:
```
[the remark this answers is not in the document]
Comment (higherstar) [reply src/RestApi/Auth.py:84]: I'm using current_identity_roles() method like this

`if "admin" in current_identity_roles():`

But when I display privileges it's looking like this;

`"privileges": [ "ADD_ROLE", "VIEW_ROLE", "EDIT_ROLE", "DELETE_ROLE", "ADD_USER", "EDIT_USER", "VIEW_USER", "EXPORT_USER", "ACTIVATE_USER", "DEACTIVATE_USER", "ARCHIVE_USER", "EDIT_USER_WORKTIME", "EDIT_USER_VACATION" ]`

So I think it's not matching with my admin control? This project was using 2 roles before. Admin and user, I changed the project like this for Keycloak. For example; When we create tm, it was like this before;

`decorators = [PermissionChecker(user_permission)]`

`if current_identity.role != Users.ADMIN and not Tags.has_specified(tag_ids): abort(403, message="Tags should include at least one private or public tag")`

I changed like this;

`@permission("user")`

So I didn't change business.

[the remark this answers is not in the document]
Comment (DanBondarenko) [reply src/RestApi/Auth.py:84]: @higherstar Hey. While NecTM in its original form uses roles-based authorization, this is not the case for our intended use. In our system, we will solely be using privileges. This means that the existing NecTM’s solution for authorization will have to be rewritten to some extent.

The list of privileges relevant to your task is described in the [Wiki subchapter](https://github.com/keeleinstituut/tv-tolkevarav/wiki/7.-T%C3%B5lkem%C3%A4lude-haldusmoodul-(T%C3%B5lkem%C3%A4lude-genereerimine-ja-t%C3%B5lkem%C3%A4lude-allalaadimise-v%C3%B5imalus)#1-requirements) linked in the [GitHub issue assigned to you](https://github.com/keeleinstituut/tv-tolkevarav/issues/232).

The current most up-to-date list of ALL privileges is (but might change during development):
* `ADD_ROLE`
* `VIEW_ROLE`
* `EDIT_ROLE`
* `DELETE_ROLE`
* `ADD_USER`
* `EDIT_USER`
* `VIEW_USER`
* `EXPORT_USER`
* `ACTIVATE_USER`
* `DEACTIVATE_USER`
* `ARCHIVE_USER`
* `EDIT_USER_WORKTIME`
* `EDIT_USER_VACATION`
* `ADD_DEPARTMENT`
* `EDIT_DEPARTMENT`
* `DELETE_DEPARTMENT`
* `EDIT_INSTITUTION`
* `EDIT_INSTITUTION_WORKTIME`
* `ADD_TAG`
* `EDIT_TAG`
* `DELETE_TAG`
* `VIEW_VENDOR_DB`
* `EDIT_VENDOR_DB`
* `VIEW_GENERAL_PRICELIST`

**Pull request closed without merging**